examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,272 @@
+    update_occupied_heap_size_on_addition(random_size * mem::size_of::<usize>());
+}
+
+fn worker_thread(tid: i32, pair_clone: Arc<(Mutex<bool>, Condvar)>, MAGIC_NUM_TO_CHECK: usize) {
+    /* Wait for all the threads to be created and then start together */
+    wait_per_thread(pair_clone);
+
+    let mut count = 0;
+    let mut tot_time_ns = 0;
+
+    let mut array_of_vectors: Vec<Vec<usize>> = Vec::new();
+
+    /* Once the thread's allocation and deallocation operations begin, we
+     * shouldn't take any lock as the allocator that we trying to test is a
+     * multithreaded allocator and we should allow as many threads as possible
+     * to get the lock.
+     */
+    for i in 1..=MAX_ITERATIONS_PER_THREAD {
+        let ran_choice = get_random_num(0, NUM_OPERATION_CHOICES - 1);
+
+        match (ran_choice) {
+            0 => {
+                select_and_check_random_buffer_contents(&array_of_vectors, MAGIC_NUM_TO_CHECK);
+                println!("T-{} check", tid);
+            }
+            1..=2 => {
+                /* Although  get_free_heap_percentage() is thread safe, this
+                 * match case may not be completely thread safe as we are only
+                 * interested in an approximate value of the remaining free space
+                 * percentage.
+                 */
+                if (get_free_heap_percentage() > MAX_ALLOWED_FREE_HEAP_PERCENTAGE) {
+                    add_new_buffer(&mut array_of_vectors, MAGIC_NUM_TO_CHECK);
+                    println!("T-{} allocate", tid);
+                } else {
+                    println!("T-{} SKIP", tid);
+                }
+            }
+            3 => {
+                delete_random_buffer(&mut array_of_vectors);
+                println!("T-{} delete", tid);
+            }
+            _ => {
+                panic!("Invalid random operation choice done");
+            }
+        }
+    }
+}
+
+fn spawn_threads(thread_count: i32, MAGIC_NUM_TO_CHECK: usize) {
+    let mut handles = vec![];
+
+    let pair = Arc::new((Mutex::new(false), Condvar::new()));
+    for i in 0..thread_count {
+        // Spawn a thread that waits until the condition is met
+        let pair_clone = Arc::clone(&pair);
+        let handle = thread::spawn(move || {
+            worker_thread(i, pair_clone, MAGIC_NUM_TO_CHECK);
+        });
+        handles.push(handle);
+    }
+
+    /* Start all the threads */
+    wakeup_all_child_threads(pair);
+
+    /* Wait for all threads to finish */
+    for handle in handles {
+        handle.join().unwrap();
+    }
+}
+
+fn get_num_processors() -> usize {
+    //num_cpus::get()
+    /* ToDo: Currently it tests with a hardcoded value. We need to add a
+     * special service to make it work properly.
+     */
+    NUM_CPUS
+}
+
+/* If there are n processors available, we will record the numbers with,2n threads,
+ * then n threads, then n/2 and so on.
+ */
+fn start_tests() {
+    let MAGIC_NUM_TO_CHECK = get_random_num(1, usize::MAX);
```

Comment:
My intention with better randomization of the fill data was that every element of every vec should get a distinct value. I think that's fine as a future enhancement.